### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by Hugobros3**
You didn't work in a seperate branch or a fork tagged as such, so I made my own to use Github as an annotation tool

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{
```
Comment (Hugobros3) [inline .vscode/launch.json]: remember to nuke this personal stuff at some point, doesn't belong in the final repo

File: src/liboslcomp/ast.cpp [hunk 2]
```diff
@@ -1108,6 +1164,7 @@ ASTassign_expression::opword() const
 
 
 
+
```
Comment (Hugobros3) [inline src/liboslcomp/ast.cpp]: I see a lot of those random spaces... you should be more careful with diffs

File: src/liboslcomp/ast.h [hunk 3]
```diff
@@ -143,6 +145,8 @@ class ASTNode : public OIIO::RefCnt {
     /// course).
     virtual Symbol* codegen(Symbol* dest = NULL);
 
+    virtual Symbol* codegen_artic(Symbol* dest = NULL);
```
Comment (Hugobros3) [inline src/liboslcomp/ast.h]: this is bad but we already talked about it ... it's still bad though. if we get serious with this after the seminar, we should do it in a better way

File: src/liboslcomp/codegen.cpp [hunk 4]
```diff
@@ -370,7 +385,6 @@ ASTNode::codegen_int(Symbol*, bool boolify, bool invert)
 }
 
 
-
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: more random formatting changes...

File: src/liboslcomp/codegen.cpp [hunk 5]
```diff
@@ -381,15 +395,81 @@ ASTshader_declaration::codegen(Symbol*)
...
     codegen_list(statements());
     return NULL;
 }
 
+Symbol*
+ASTshader_declaration::codegen_artic(Symbol*)
+{
+    auto name = this->shadername();
+
+
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: what's up with those 3-newline spaces ? is that a code style thing from the OSL repo ?

File: src/liboslcomp/codegen.cpp [hunk 6]
```diff
@@ -381,15 +395,81 @@ ASTshader_declaration::codegen(Symbol*)
...
+
+    std::cout << "struct " << name << "_in {" << std::endl;
+    //NOT_IMPLEMENTED;
+
+    std::vector<ref> outputs = {};
+    for (ref f = formals(); f; f = f->next()) {
+        OSL_DASSERT(f->nodetype() == ASTNode::variable_declaration_node);
+        ASTvariable_declaration* v = (ASTvariable_declaration*)f.get();
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: should probably be called "param"

File: src/liboslcomp/codegen.cpp [hunk 7]
```diff
@@ -381,15 +395,81 @@ ASTshader_declaration::codegen(Symbol*)
...
+        }
+        std::cout << "," << std::endl;
+    }
+
+    std::cout << "}" << std::endl
+              << std::endl
+              << "struct " << name << "_out {" << std::endl;
+    for (auto&& f : outputs) {
+        OSL_DASSERT(f->nodetype() == ASTNode::variable_declaration_node);
+        ASTvariable_declaration* v = (ASTvariable_declaration*)f.get();
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: ditto

File: src/liboslcomp/codegen.cpp [hunk 8]
```diff
@@ -535,6 +615,81 @@ ASTassign_expression::codegen(Symbol* dest)
     return dest;
 }
 
+Symbol* ASTassign_expression::codegen_artic(Symbol *dest){
+
+    OSL_DASSERT(m_op == Assign);  // all else handled by binary_op
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: tip: you can use ASSERT_MACRO(condition && "message") for those

File: src/liboslcomp/codegen.cpp [hunk 9]
```diff
@@ -535,6 +615,81 @@ ASTassign_expression::codegen(Symbol* dest)
...
+
+    ASTindex* index = NULL;
+    if (var()->nodetype() == index_node) {
+        NOT_IMPLEMENTED;
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: keep NOT_IMPLEMENTED macros at the end of unimplemented sections

File: src/liboslcomp/codegen.cpp [hunk 10]
```diff
@@ -535,6 +615,81 @@ ASTassign_expression::codegen(Symbol* dest)
...
+        // named component (P.x = ...).
+    } else {
+        dest = var()->codegen_artic();
+    }
+    std::cout << " = ";
+    Symbol* operand = expr()->codegen_artic(dest);
+        OSL_DASSERT(operand != NULL);
+
+    if (typespec().is_structure_array()) {
+        // Assign entire array-of-struct to another array-of-struct
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: what is special about arrays of structs ? you can assign those fine in Artic

File: src/liboslcomp/codegen.cpp [hunk 11]
```diff
@@ -535,6 +615,81 @@ ASTassign_expression::codegen(Symbol* dest)
...
+            StructSpec* structspec = typespec().structspec();
+            codegen_assign_struct_artic(structspec, ustring(dest->mangled()),
+                                  ustring(operand->mangled()), NULL, true, 0,
+                                  false /*not a shader param init*/);
+        }
+        return dest;
+    }
+
+    if (typespec().is_structure()) {
+        // Assignment of struct copies each element individually
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: again something Artic handles natively... are you sure this is necessary ?

File: src/liboslcomp/codegen.cpp [hunk 12]
```diff
@@ -535,6 +615,81 @@ ASTassign_expression::codegen(Symbol* dest)
...
+        ASTstructselect* ss = (ASTstructselect*)var().get();
+        ss->codegen_assign_artic(dest, operand);
+        return dest;
+    }
+
+    if (index) {
+        index->codegen_assign_artic(operand);
+        dest = operand;  // so transitive assignment works for array refs
+    } else if (operand != dest) {
+        emitcode(typespec().is_array() ? "arraycopy" : "assign", dest, operand);
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: using strings for this is awful, i know it's not your fault but i still hate it :(

File: src/liboslcomp/codegen.cpp [hunk 13]
```diff
@@ -755,6 +910,177 @@ ASTNode::one_default_literal(const Symbol* sym, ASTNode* init, std::string& out,
...
+    ASTliteral* lit = static_cast<ASTliteral*>(init);
+    bool completed  = true;  // have we output the full initialization?
+    TypeSpec type   = sym->typespec().elementtype();
+    if (type.is_closure()) {
+        // this clause avoid trouble and assertions if the following
+        // is_int(), i_float(), etc, encounter a closure.
+        completed = islit;
+    } else if (type.is_structure()) {
+        // No initializers for struct
+        completed = false;
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: why not return false immediately ?

File: src/liboslcomp/codegen.cpp [hunk 14]
```diff
@@ -855,6 +1240,51 @@ ASTvariable_declaration::codegen_initializer(ref init, Symbol* sym)
...
+                if (paraminit)
+                    sym->initend(m_compiler->next_op_label());
+                return;
+            }
+        }
+        codegen_initlist_artic(init, m_typespec, m_sym);
+    }
+
+}
+
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: IMO you also should not have copy-pasted those methods, it's once again very hard to tell what you actually wrote

File: src/liboslcomp/codegen.cpp [hunk 15]
```diff
@@ -1831,7 +2643,7 @@ ASTfunction_call::codegen(Symbol* dest)
         TypeSpec formaltype = (i < (int)polyargs.size())
                                   ? polyargs[i]
                                   : TypeSpec(TypeDesc::UNKNOWN);
-        bool writearg = argwrite(i + returnarg);
+        bool writearg       = argwrite(i + returnarg);
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: wat

File: src/liboslexec/typespec.cpp [hunk 16]
```diff
@@ -91,65 +93,155 @@ TypeSpec::structure_id (const char *name, bool add)
...
+TypeSpec::artic_string()
+{
+    if (this->is_array()) {
+        std::string start = "";
+        std ::string end  = "";
+        start += "[";
+
+        if (this->is_closure_array()) {
+            if(this->is_color_closure()){
+                start += "Closure<Color<f32>>";
```
Comment (Hugobros3) [inline src/liboslexec/typespec.cpp]: Artic does not use this syntax, but you probably found that out already

File: src/liboslexec/typespec.cpp [hunk 17]
```diff
@@ -91,65 +93,155 @@ TypeSpec::structure_id (const char *name, bool add)
...
+TypeSpec::artic_simpletype()
+{
+    TypeDesc st       = this->simpletype();
+    if(st.is_unknown()){
+        return "f32";
+    }
+    std::string start = "";
+    std::string end   = "";
+    if (st.is_array()) {
+        start += "[";
```
Comment (Hugobros3) [inline src/liboslexec/typespec.cpp]: missing something here aren't you

File: src/liboslcomp/codegen.cpp [hunk 18]
```diff
@@ -949,25 +951,24 @@ ASTNode::one_default_literal_artic(const Symbol* sym, ASTNode* init,
         }
     } else if (type.is_triple()) {
         if (type.is_point()) {
-            std::cout << "Point<";
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: why std::cout and not the std::string& "out" param ?

File: src/liboslcomp/codegen.cpp [hunk 19]
```diff
@@ -2562,6 +2572,81 @@ ASTtype_constructor::codegen(Symbol* dest)
...
+    } else if (nargs == 1) {
+        emitcode("assign", argdest.size(),
+                 (argdest.size()) ? &argdest[0] : NULL);
+    } else
+        emitcode(typespec().string().c_str(), argdest.size(),
+                 (argdest.size()) ? &argdest[0] : NULL);
+    std::cout << "}";
+    return dest;
+}
+
```
Comment (Hugobros3) [inline src/liboslcomp/codegen.cpp]: more copy/paste of the original methods...

File: src/liboslcomp/artic.cpp [hunk 20]
```diff
@@ -0,0 +1,84 @@
+//
+// Created by misha on 14/06/2021.
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: who is misha ?

File: src/liboslcomp/artic.cpp [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+//
+#include "artic.h"
+
+void
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: yes, this is seemingly their official codestyle, but you have my blessing to stray away (far away) from it, you already changed the formatting in many places accidentally

File: src/liboslcomp/ast.h [hunk 22]
```diff
@@ -145,7 +146,7 @@ class ASTNode : public OIIO::RefCnt {
     /// course).
     virtual Symbol* codegen(Symbol* dest = NULL);
```
Comment (Hugobros3) [inline src/liboslcomp/ast.h]: honest question: why does codegen _return_ a symbol ?

File: src/liboslcomp/artic.cpp [hunk 23]
```diff
@@ -3,10 +3,147 @@
...
+    }
+}
+
+std::string
+artic_type_string_to_string(TypeSpec typeSpec)
+{
+    auto tstring = artic_string(typeSpec);
+    size_t pos;
+    while ((pos = tstring.find('[')) != std::string::npos) {
+        tstring.replace(pos, 1, "_");
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: wait seriously ?

File: src/liboslcomp/artic.cpp [hunk 24]
```diff
@@ -3,10 +3,147 @@
...
+    }
+
+    while ((pos = tstring.find('<')) != std::string::npos) {
+        tstring.replace(pos, 1, "_");
+    }
+
+    while ((pos = tstring.find('>')) != std::string::npos) {
+        tstring.replace(pos, 1, "_");
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: is this how you do mangling ?

File: src/liboslcomp/artic.cpp [hunk 25]
```diff
@@ -67,18 +204,23 @@ ArticSource::save_temp(std::string in)
...
 {
     return std::move(m_buffer);
 }
 ArticSource
 ArticSource::make_temp_source()
 {
-    auto is = this->m_indent_string;
-    auto a = ArticSource(is);
+    auto is    = this->m_indent_string;
+    auto a     = ArticSource(is);
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: i don't think aligning this is worth it

File: src/liboslcomp/artic.cpp [hunk 26]
```diff
@@ -223,4 +234,109 @@ ArticSource::integrate_temp_source(ArticSource&& articSource)
...
 }
 
+void ArticTranspiler::dispatch_node(ASTNode::ref n)
+{
+    auto node = n.get();
+    switch (node->nodetype()) {
+    case ASTNode::unknown_node:
+        NOT_IMPLEMENTED;
+        break;
+    case ASTNode::shader_declaration_node:
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: there is a nice macro trick we use in thorin to avoid this kind of thing, i can explain it to you in a call at some point

File: src/liboslcomp/oslcomp.cpp [hunk 27]
```diff
@@ -466,9 +467,12 @@ OSLCompilerImpl::compile(string_view filename,
             write_dependency_file(filename);
 
         if (!error_encountered()) {
-            //ArticSource artic_source("  ");
-            //shader()->codegen_artic(artic_source);
-            //artic_source.print();
+
+            ArticSource artic_source("  ");
+            ArticTranspiler artic_transpiler(artic_source);
+            artic_transpiler.dispatch_node(shader());
```
Comment (Hugobros3) [inline src/liboslcomp/oslcomp.cpp]: does that make the call to codegen() below redundant ?

File: src/liboslcomp/artic.cpp [hunk 28]
```diff
@@ -474,17 +472,94 @@ ArticTranspiler::transpile_typecast_expression(ASTtypecast_expression* node)
...
+
 void
 ArticTranspiler::transpile_function_call(ASTfunction_call* node)
 {
-    NOT_IMPLEMENTED;
+    if(node->is_struct_ctr()){
+        auto constructor = ASTtype_constructor(nullptr, node->typespec(), node->args().get());
+        transpile_type_constructor(&constructor);
+    } else {
+        std::vector<ASTNode::ref> args = {};
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: why store those intermediate results ? can't you dispatch directly ?

File: src/liboslcomp/artic.cpp [hunk 29]
```diff
@@ -474,17 +472,94 @@ ArticTranspiler::transpile_typecast_expression(ASTtypecast_expression* node)
...
+        NOT_IMPLEMENTED;
+    }
+}
+
+
+std::string
+ArticTranspiler::get_arg_name(TypeSpec typeSpec, int argnum)
+{
+    if (typeSpec.is_triple() && !typeSpec.is_color()) {
+        switch (argnum) {
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: `std::string(1, "xyz"[i])`

File: src/liboslcomp/artic.cpp [hunk 30]
```diff
@@ -474,17 +472,94 @@ ArticTranspiler::transpile_typecast_expression(ASTtypecast_expression* node)
...
 ArticTranspiler::transpile_literal_node(ASTliteral* node)
 {
-    NOT_IMPLEMENTED;
+    if(node->typespec().is_int()){
+        source.add_source(std::to_string(node->intval()));
+    } else if(node->typespec().is_float()){
+        source.add_source(std::to_string(node->floatval()));
+    } else if(node->typespec().is_string()){
+        add_string_constant(node->strval());
+        source.add_source("Strings::", node->strval());
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: ?

File: src/liboslcomp/ast.h [hunk 31]
```diff
@@ -748,6 +752,164 @@ class ASTreturn_statement final : public ASTNode {
 };
 
 
+class ASTfunction_call final : public ASTNode {
```
Comment (Hugobros3) [inline src/liboslcomp/ast.h]: Why did you move this ?

File: src/liboslcomp/artic.cpp [hunk 32]
```diff
@@ -652,6 +698,26 @@ ArticTranspiler::get_arg_name(TypeSpec typeSpec, int argnum)
         OSL_ASSERT(argnum < structSpec->numfields());
         auto fieldSpec = structSpec->field(static_cast<int>(argnum));
         return fieldSpec.name.string();
+    } else if(typeSpec.is_matrix()){
+        switch (argnum){
+        case 0: return "m1_n1";
```
Comment (Hugobros3) [inline src/liboslcomp/artic.cpp]: please just generate this

**Review by Hugobros3 — COMMENTED** [review]
Haven't checked over -everything-, but covered plenty of things already

File: src/liboslcomp/ast.h (hunk 3 above)
Comment (JoschuaL) [reply src/liboslcomp/ast.h]: the whole codegen_artic thing is already nuked and replaced with a proper ast visitor

File: src/liboslcomp/artic.cpp (hunk 28 above)
Comment (JoschuaL) [reply src/liboslcomp/artic.cpp]: If I do that I get a pointer exception because of the weird Ref counting the OSL SDk is doing internally with all AST nodes, and there doesnt seem to be a way to opt out of it

File: src/liboslcomp/artic.cpp (hunk 30 above)
Comment (JoschuaL) [reply src/liboslcomp/artic.cpp]: this is for converting constant strings into enums, so that all the functions that just use strings as pseudo-enums (like transforming from one coordinate space to another) without actually having to deal with strings

File: src/liboslcomp/artic.cpp (hunk 29 above)
Comment (JoschuaL) [reply src/liboslcomp/artic.cpp]: that doesn't cover i > 2 though. Which might be better, but I'm not quite sure

File: src/liboslcomp/ast.h (hunk 31 above)
Comment (JoschuaL) [reply src/liboslcomp/ast.h]: if I remember correctly so I could add a constructor to ASTtype_constructor that takes a ASTfunction_call (for cases when the lexer interprets a constructor as a function call, which happens quite frequently). To do so I had to move the definition of ASTfunction_call before ASTtype_constructor. There might be a way around that, but forward declaration did not work, and I'm just not really all that familiar with the whole C++ name resolution trickery.

File: src/liboslcomp/artic.cpp (hunk 32 above)
Comment (JoschuaL) [reply src/liboslcomp/artic.cpp]: fair enough xD

File: src/liboslcomp/artic.cpp (hunk 28 above)
Comment (JoschuaL) [reply src/liboslcomp/artic.cpp]: never mind, you cant take the address of temporary objects, and I dont want to implement another constructor just in order to get a pointer (so writing
`transpile_type_constructor(&ASTtype_constructor(node));`
doesn't compile)
